daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -1552,18 +1552,38 @@ static krb5_error_code dbget_alias(krb5_context kcontext,
     krb5_db_entry *kentry = NULL;
     krb5_data *realm;
     krb5_boolean check = FALSE;
+    krb5_int32 supported_types[] = {
+        KRB5_NT_ENTERPRISE_PRINCIPAL,
+        KRB5_NT_PRINCIPAL,
+        -1,
+    };
+    size_t i = 0;
 
-    /* TODO: also support hostbased aliases */
+    /* For TGS-REQ server principal lookup, KDC asks with KRB5_KDB_FLAG_REFERRAL_OK
+     * and client usually asks for an KRB5_NT_PRINCIPAL type principal. */
+    if ((flags & KRB5_KDB_FLAG_REFERRAL_OK) == 0) {
+       /* this is *not* TGS-REQ server principal search, remove
+	* KRB5_NT_PRINCIPAL from the supported principal types for this lookup */
+       supported_types[1] = -1;
+    }
 
     /* Enterprise principal name type is for potential aliases or principals
-     * from trusted realms. The logic below only applies to this type */
-    if (krb5_princ_type(kcontext, search_for) != KRB5_NT_ENTERPRISE_PRINCIPAL) {
+     * from trusted realms. Except for the TGS-REQ server lookup, we only
+     * expect enterprise principals here */
+    for (i = 0; supported_types[i] != -1; i++) {
```

Comment:
Future additions to the `supported_types` array might lead to bugs because inserting a new type will break the reference by rank at line 1567, and appending a new type would cause it to be removed altogether with `KRB5_NT_PRINCIPAL` at the same line.

I would be in favor of defining the array this way:
```c
krb5_int32 supported_types[] = {
    [0] = KRB5_NT_ENTERPRISE_PRINCIPAL,
    [1] = KRB5_NT_PRINCIPAL,
    -1,
}
```
It emphasize the fact ordering matters, which will hopefully help future contributors to notice modifying it also requires updating the `KRB5_NT_PRINCIPAL` rank reference.

Reply:
I applied this logic and also added few comments to clarify the code.